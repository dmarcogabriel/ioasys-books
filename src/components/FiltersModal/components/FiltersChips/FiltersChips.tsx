import React from 'react';
import {
  FiltersChipsContainer,
  FiltersChipsTitle,
  FiltersChipsOptions,
  ChipButton,
  ChipButtonText,
} from './styles';

export interface Option {
  name: string;
  value: string;
}

export interface FiltersChipsProps {
  title: string;
  options: Option[];
  value: string | null;
  onChange: (option: string) => void;
  testID?: string;
}

export const FiltersChips = ({
  title,
  options,
  value,
  testID = '',
  onChange,
}: FiltersChipsProps): JSX.Element => {
  const isSelected = (option: string): boolean => {
    return !!value && value === option;
  };

  const handleSelect = (option: string): void => {
    onChange(option);
  };

  return (
    <FiltersChipsContainer>
      <FiltersChipsTitle testID="filtersChipsTitle">{title}</FiltersChipsTitle>
      <FiltersChipsOptions testID="filtersChipsOptions">
        {options.map((option, i) => (
          <ChipButton
            key={`${i}`}
            testID={`filtersChipsOptionButton${testID}${i + 1}`}
            isSelected={isSelected(option.value)}
            onPress={() => handleSelect(option.value)}>
            <ChipButtonText
              testID={`filtersChipsOptionText${testID}${i + 1}`}
              isSelected={isSelected(option.value)}>
              {option.name}
            </ChipButtonText>
          </ChipButton>
        ))}
      </FiltersChipsOptions>
    </FiltersChipsContainer>
  );
};
