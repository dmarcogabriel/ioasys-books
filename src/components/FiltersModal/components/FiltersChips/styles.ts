import styled, {css} from 'styled-components/native';
import {ThemeProps} from '../../../../@types/theme.interface';

export const FiltersChipsContainer = styled.View`
  margin: 16px 0;
`;

export const FiltersChipsTitle = styled.Text`
  color: ${({theme}: ThemeProps) => theme.colors.dark};
  font-size: 12px;
  font-weight: bold;
  margin: 4px 0;
`;

export const FiltersChipsOptions = styled.View`
  flex-direction: row;
  margin: 4px 0;
`;

interface ChipButtonProps {
  isSelected: boolean;
}

export const ChipButton = styled.TouchableOpacity`
  margin: 0 4px;
  ${({theme, isSelected}: ThemeProps & ChipButtonProps) => css`
    background: ${isSelected ? theme.colors.dark : theme.colors.light};
    border: 1px solid ${theme.colors.dark};
  `};
  border-radius: 20px;
  padding: 8px 20px;
  align-items: center;
  justify-content: center;
`;

export const ChipButtonText = styled.Text`
  ${({theme, isSelected}: ThemeProps & ChipButtonProps) => css`
    color: ${isSelected ? theme.colors.light : theme.colors.dark};
  `};
  font-size: 12px;
`;
