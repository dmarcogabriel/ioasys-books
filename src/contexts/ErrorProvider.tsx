import React, {createContext} from 'react';
import {UNAUTHORIZED_STATUS} from '../constants';
import Toast from 'react-native-toast-message';

type ErrorData = {
  message: string;
};

export interface ApiError {
  errors: ErrorData;
}

interface ErrorContextValue {
  handleError: (
    error: ApiError,
    status: number,
    logoutCallback: () => Promise<void>,
  ) => Promise<void>;
}

export const ErrorContext = createContext<ErrorContextValue>({
  handleError: async () => {},
});

interface ErrorProviderProps {
  children: React.ReactNode;
}

export const ErrorProvider = ({children}: ErrorProviderProps): JSX.Element => {
  const handleError = async (
    error: ApiError,
    status: number,
    logoutCallback: () => Promise<void>,
  ): Promise<void> => {
    let errorMessage: string = '';
    if (status === UNAUTHORIZED_STATUS) {
      await logoutCallback();
      errorMessage = 'Sess√£o expirada';
    } else {
      Toast.show({
        position: 'bottom',
        text1: errorMessage || error.errors.message,
        type: 'error',
      });
    }
  };

  return (
    <ErrorContext.Provider value={{handleError}}>
      {children}
    </ErrorContext.Provider>
  );
};
