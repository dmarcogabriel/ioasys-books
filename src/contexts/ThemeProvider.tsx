import React from 'react';
import {ThemeProvider as SCThemeProvider} from 'styled-components/native';

interface ThemeProviderProps {
  children: React.ReactNode;
}

export const ThemeProvider = ({children}: ThemeProviderProps): JSX.Element => {
  const theme = {
    colors: {
      primary: '#B22E6F',
      dark: '#333333',
      light: '#ffffff',
      grey: '#999999',
    },
  };

  return <SCThemeProvider theme={theme}>{children}</SCThemeProvider>;
};
