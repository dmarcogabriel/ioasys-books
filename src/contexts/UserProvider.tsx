import React, {createContext, useState, useEffect, useCallback} from 'react';
import {User} from '../@types/user.interface';
import {storeData, removeData, getData} from '../utils/storage';
import {AUTH_TOKEN_STORAGE_KEY} from '../constants';

interface UserContextValue {
  user: User | null;
  login: (user: User, token: string) => Promise<void>;
  logout: () => Promise<void>;
  token: string | null;
}

export const UserContext = createContext<UserContextValue>({
  user: null,
  token: null,
  login: async () => {},
  logout: async () => {},
});

interface UserProviderProps {
  children: React.ReactNode;
}

export const UserProvider = ({children}: UserProviderProps): JSX.Element => {
  const [user, setUser] = useState<User | null>(null);
  const [token, setToken] = useState<string | null>(null);
  const [isLoading, setIsLoading] = useState(true);

  const login = async (data: User, authToken: string): Promise<void> => {
    setUser(data);
    setToken(authToken);
    await storeData(AUTH_TOKEN_STORAGE_KEY, authToken);
  };

  const logout = async () => {
    await removeData(AUTH_TOKEN_STORAGE_KEY);
    setUser(null);
    setToken(null);
  };

  const loadToken = useCallback(async () => {
    setIsLoading(true);
    try {
      const authToken = await getData<string>(AUTH_TOKEN_STORAGE_KEY);
      if (authToken) {
        setToken(authToken);
      }
    } catch (error) {
      setIsLoading(false);
    }
    setIsLoading(false);
  }, []);

  useEffect(() => {
    loadToken();
  }, [loadToken]);

  return (
    <UserContext.Provider value={{user, token, login, logout}}>
      {!isLoading && children}
    </UserContext.Provider>
  );
};
