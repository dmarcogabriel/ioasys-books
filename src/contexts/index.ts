export * from './UserProvider';
export * from './ThemeProvider';
export * from './ErrorProvider';
