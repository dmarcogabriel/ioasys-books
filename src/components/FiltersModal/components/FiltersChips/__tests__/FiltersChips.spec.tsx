import React from 'react';
import {render, fireEvent} from '@testing-library/react-native';
import {FiltersChips, FiltersChipsProps, Option} from '../FiltersChips';
import {ThemeProvider} from '../../../../../contexts';
import renderer from 'react-test-renderer';

const mockNavigate = jest.fn();
jest.mock('@react-navigation/native', () => ({
  useNavigation: () => ({
    navigate: mockNavigate,
  }),
}));

const ComponentWrapper = (props: FiltersChipsProps) => (
  <ThemeProvider>
    <FiltersChips {...props} />
  </ThemeProvider>
);

let mockSelectedOption: string | null;
const mockOnChange = jest.fn((option: string | null) => {
  mockSelectedOption = option;
});
const mockTitle = 'Test';
const mockOptions = ['option 1', 'option 2', 'option 3'];

const mockOptionsMapper = (options: string[]): Option[] => {
  return options.map(option => ({
    name: option,
    value: option,
  }));
};

describe('components/FiltersModal/components/FiltersChips', () => {
  it('should render correctly', () => {
    const {getByTestId} = render(
      <ComponentWrapper
        title={mockTitle}
        options={mockOptionsMapper(mockOptions)}
        onChange={mockOnChange}
        value=""
      />,
    );
    const filtersChipsTitle = getByTestId('filtersChipsTitle');
    const filtersChipsOptions = getByTestId('filtersChipsOptions');
    const filtersChipsOptionText1 = getByTestId('filtersChipsOptionText1');
    const filtersChipsOptionText2 = getByTestId('filtersChipsOptionText2');
    const filtersChipsOptionText3 = getByTestId('filtersChipsOptionText3');

    expect(filtersChipsTitle.children[0]).toBe('Test');
    expect(filtersChipsOptions.children).toHaveLength(3);
    expect(filtersChipsOptionText1.children[0]).toBe('option 1');
    expect(filtersChipsOptionText2.children[0]).toBe('option 2');
    expect(filtersChipsOptionText3.children[0]).toBe('option 3');
  });

  it('should render correctly with selected option', () => {
    const {getByTestId} = render(
      <ComponentWrapper
        value="option 2"
        title={mockTitle}
        options={mockOptionsMapper(mockOptions)}
        onChange={mockOnChange}
      />,
    );
    const filtersChipsOptionText1 = getByTestId('filtersChipsOptionText1');
    const filtersChipsOptionText2 = getByTestId('filtersChipsOptionText2');
    const filtersChipsOptionText3 = getByTestId('filtersChipsOptionText3');

    expect(filtersChipsOptionText1).toHaveProp('isSelected', false);
    expect(filtersChipsOptionText2).toHaveProp('isSelected', true);
    expect(filtersChipsOptionText3).toHaveProp('isSelected', false);
  });

  it('should pass on select option', () => {
    const {getByTestId} = render(
      <ComponentWrapper
        value=""
        title={mockTitle}
        options={mockOptionsMapper(mockOptions)}
        onChange={mockOnChange}
      />,
    );
    fireEvent.press(getByTestId('filtersChipsOptionButton2'));

    expect(mockOnChange).toHaveBeenCalled();
    expect(mockSelectedOption).toBe('option 2');
  });

  it('should match snapshot', () => {
    expect(
      renderer
        .create(
          <ComponentWrapper
            title={mockTitle}
            options={mockOptionsMapper(mockOptions)}
            value="option 1"
            onChange={mockOnChange}
          />,
        )
        .toJSON(),
    ).toMatchSnapshot();
  });
});
