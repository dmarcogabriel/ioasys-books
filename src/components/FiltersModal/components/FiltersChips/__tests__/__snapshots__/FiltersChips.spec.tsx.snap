// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/FiltersModal/components/FiltersChips should match snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "marginBottom": 16,
        "marginLeft": 0,
        "marginRight": 0,
        "marginTop": 16,
      },
    ]
  }
>
  <Text
    style={
      Array [
        Object {
          "color": "#333333",
          "fontSize": 12,
          "fontWeight": "bold",
          "marginBottom": 4,
          "marginLeft": 0,
          "marginRight": 0,
          "marginTop": 4,
        },
      ]
    }
    testID="filtersChipsTitle"
  >
    Test
  </Text>
  <View
    style={
      Array [
        Object {
          "flexDirection": "row",
          "flexWrap": "wrap",
          "marginBottom": 4,
          "marginLeft": 0,
          "marginRight": 0,
          "marginTop": 4,
        },
      ]
    }
    testID="filtersChipsOptions"
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      nativeID="animatedComponent"
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#333333",
          "borderColor": "#333333",
          "borderRadius": 20,
          "borderStyle": "solid",
          "borderWidth": 1,
          "justifyContent": "center",
          "marginBottom": 4,
          "marginLeft": 4,
          "marginRight": 4,
          "marginTop": 4,
          "opacity": 1,
          "paddingBottom": 8,
          "paddingLeft": 20,
          "paddingRight": 20,
          "paddingTop": 8,
        }
      }
      testID="filtersChipsOptionButton1"
    >
      <Text
        isSelected={true}
        style={
          Array [
            Object {
              "color": "#ffffff",
              "fontSize": 12,
            },
          ]
        }
        testID="filtersChipsOptionText1"
      >
        option 1
      </Text>
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      nativeID="animatedComponent"
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#ffffff",
          "borderColor": "#333333",
          "borderRadius": 20,
          "borderStyle": "solid",
          "borderWidth": 1,
          "justifyContent": "center",
          "marginBottom": 4,
          "marginLeft": 4,
          "marginRight": 4,
          "marginTop": 4,
          "opacity": 1,
          "paddingBottom": 8,
          "paddingLeft": 20,
          "paddingRight": 20,
          "paddingTop": 8,
        }
      }
      testID="filtersChipsOptionButton2"
    >
      <Text
        isSelected={false}
        style={
          Array [
            Object {
              "color": "#333333",
              "fontSize": 12,
            },
          ]
        }
        testID="filtersChipsOptionText2"
      >
        option 2
      </Text>
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      nativeID="animatedComponent"
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#ffffff",
          "borderColor": "#333333",
          "borderRadius": 20,
          "borderStyle": "solid",
          "borderWidth": 1,
          "justifyContent": "center",
          "marginBottom": 4,
          "marginLeft": 4,
          "marginRight": 4,
          "marginTop": 4,
          "opacity": 1,
          "paddingBottom": 8,
          "paddingLeft": 20,
          "paddingRight": 20,
          "paddingTop": 8,
        }
      }
      testID="filtersChipsOptionButton3"
    >
      <Text
        isSelected={false}
        style={
          Array [
            Object {
              "color": "#333333",
              "fontSize": 12,
            },
          ]
        }
        testID="filtersChipsOptionText3"
      >
        option 3
      </Text>
    </View>
  </View>
</View>
`;
